ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala

```diff
@@ -185,7 +183,10 @@ private[events] object FilterTableACSReader {
     val queueState = new QueueState(signalQueue, initialTasks)
 
     signalSource
-      .mapAsyncUnordered(workerParallelism) { _ =>
+      // in theory signals could be executed parallel without maintaining order with mapAsyncUnordered,
+      // but in practice with mapAsync we force the parallel processing to "wait" for a slow work item,
+      // which prevents excessive memory consumption in later stages.
```

Comment:
I find it quite hard to verify the correctness of such claims in the current ACSReader code, simply because the code is so complex.

Does this really guarantee that all tasks progress at the same rate, especially for `workerParallelism>2`? Consider the following case:
- You have two tasks F and S. F is fast, S is slow.
- You have workerParallelism of 3.
- For each batch emitted by S, two batches are emitted by F. This will lead to a pattern `FFSFFSFFS` in the queue.
Will this not lead to data from the fast task piling up in the id merging stage?

Reply:
It does not guarantee it, but it neither aims to do so. It only guarantees that a slow Future is slowing down the signal pipeline.
* Due to the query strategy (continuous index range scans) the fetching duration depend mostly on the id-page-size: in practice these will be equally fast
* Rate in the problem statement could mean a dense and a sparse range too: since the next task is picked from a priority-queue (the smallest), and since in practice the second phase fetching is slower than the first one (therefore backpressures), demand is low, which boils down to low parallelism in fetching (this is supported by observation with experiments)
* On the other hand if for some reason the second phase would be faster, there would be maximum parallelism applied at id-fetching, which would lead to overfetching "slow lanes", leading to accumulation there (this was also reproducible with experiments) -- although this is an independent issue from the original (slow Future causing accumulation), with very limited impact

I agree with your observation that this logic is quite complex, and not easy to reason about -- for what it is worth the problem to solve is neither simple.
Therefore we will go ahead and try to make this machinery simpler, and more robust in terms of maximum memory consumption in the running improvement epic.